File: Gambler_Ruin_Problem_Bitcoin/gamblerRuinProblem_Bitcoin.c
#include <stdio.h>
#include <math.h>

double attackerSuccessProbability(double q, int z){
    double p = 1.0 - q;
    double lambda = z * (q/p);
    double sum = 1.0;
    int i,k;

    for(k=0;k<=z;k++){
        double poisson = exp(-lambda);
        for(i=1;i<=k;i++){
            poisson *= lambda/i;
        }
        sum -= poisson * (1-pow(q/p,z-k));
        printf("\nBlock %d -> %lf",k,sum);
    }
    return sum;
}

int main(){
    
    double q;
    int z;

    printf("Probability of the attacker finds the next block ");
    scanf("%lf",&q);
    printf("How many blocks is the attacker behind ");
    scanf("%i",&z);

    double result = attackerSuccessProbability(q,z);
    printf("\n\nResult is %lf\n", result);
    
    return 0;
}

File: Gambler_Ruin_Problem_Bitcoin/gamblerRuinProblem_Bitcoin.py
#Prove by Calculations the gambler ruin problem in the bitcoin 

#The probability of an attacker catching up from a given deficit

# p - probability an honest node finds the next block
# q - probability the attacker finds the next block
# qz - probability the attacker will ever catch up from z blocks behind

import math

q = float(input("Probability of the attacker finds the next block "))
z = int(input("How many blocks is the attacker behind "))


def attackerSuccessProbability(q,z):
    x = range(0,z+1)
    p = 1 - q
    lambda_n = z * (q/p)
    sum_n = 1

    for n in x:
        poisson = math.exp(-lambda_n)
        x_inside = range(1,n+1)
        for n_inside in x_inside:
            poisson = poisson * (lambda_n/n_inside)
        sum_n = sum_n - poisson * (1 - math.pow(q/p,z-n)) 
        print("Block nº " + str(n) + " ----> " + str(sum_n))
    return sum_n


attackerSuccessProbability(q,z)
